services/graph/pkg/service/v0/users_filter.go

```diff
@@ -32,11 +32,89 @@ func (g Graph) applyUserFilter(ctx context.Context, req *godata.GoDataRequest, r
 		return g.applyFilterLambda(ctx, req, root.Children)
 	case godata.ExpressionTokenLogical:
 		return g.applyFilterLogical(ctx, req, root)
+	case godata.ExpressionTokenFunc:
+		return g.applyFilterFunction(ctx, req, root)
 	}
 	logger.Debug().Str("filter", req.Query.Filter.RawValue).Msg("filter is not supported")
 	return users, unsupportedFilterError()
 }
 
+func (g Graph) applyFilterFunction(ctx context.Context, req *godata.GoDataRequest, root *godata.ParseNode) (users []*libregraph.User, err error) {
+	logger := g.logger.SubloggerWithRequestID(ctx)
+	if root.Token.Type != godata.ExpressionTokenFunc {
+		return users, invalidFilterError()
+	}
+
+	switch root.Token.Value {
+	case "startswith":
+		// 'startswith' needs 2 operands
+		if len(root.Children) != 2 {
+			return users, invalidFilterError()
+		}
+		return g.applyFilterFunctionStartsWith(ctx, req, root.Children[0], root.Children[1]), nil
+	case "contains":
+		// 'contains' needs 2 operands
+		if len(root.Children) != 2 {
+			return users, invalidFilterError()
+		}
+		return g.applyFilterFunctionContains(ctx, req, root.Children[0], root.Children[1]), nil
+	}
+	logger.Debug().Str("Token", root.Token.Value).Msg("unsupported function filter")
+	return users, unsupportedFilterError()
+}
+
+func (g Graph) applyFilterFunctionStartsWith(ctx context.Context, req *godata.GoDataRequest, operand1 *godata.ParseNode, operand2 *godata.ParseNode) (users []*libregraph.User) {
+	logger := g.logger.SubloggerWithRequestID(ctx)
+	if operand1.Token.Type != godata.ExpressionTokenLiteral {
+		logger.Debug().Str("Token", operand1.Token.Value).Msg("unsupported function filter")
+		return users
+	}
+
+	switch operand1.Token.Value {
+	case "displayName":
+		var retUsers []*libregraph.User
+		filterValue := operand2.Token.Value
+		logger.Debug().Str("property", operand2.Token.Value).Str("value", filterValue).Msg("Filtering displayName by startsWith")
+		if users, err := g.identityBackend.GetUsers(ctx, req); err == nil {
+			for _, user := range users {
+				if strings.HasPrefix(user.GetDisplayName(), filterValue) {
+					retUsers = append(retUsers, user)
+				}
+			}
+		}
+		return retUsers
+	default:
+		logger.Warn().Str("Token", operand1.Token.Value).Msg("unsupported function filter")
+		return nil
+	}
+}
+
+func (g Graph) applyFilterFunctionContains(ctx context.Context, req *godata.GoDataRequest, operand1 *godata.ParseNode, operand2 *godata.ParseNode) (users []*libregraph.User) {
+	logger := g.logger.SubloggerWithRequestID(ctx)
+	if operand1.Token.Type != godata.ExpressionTokenLiteral {
+		logger.Debug().Str("Token", operand1.Token.Value).Msg("unsupported function filter")
+		return users
+	}
+
+	switch operand1.Token.Value {
+	case "displayName":
+		var retUsers []*libregraph.User
+		filterValue := operand2.Token.Value
+		logger.Debug().Str("property", operand2.Token.Value).Str("value", filterValue).Msg("Filtering displayName by contains")
+		if users, err := g.identityBackend.GetUsers(ctx, req); err == nil {
```

Comment:
In theory you could use `GetUsers()` here with a patched odata req containing a `search` value. That would already to a serverside filterin (at least in the LDAP backend)

Reply:
Then let's merge it now and I see if I can squeeze the solution into another PR.

Reply:
I think i have a solution

Reply:
I wonder if it would make sense to make it more generic and just use `operand1.Token.Value` in both cases and get rid of the switch.... maybe use a string-slice whitelist (e.g. `{"displayName", "email"} etc) on what the api is allowed to filter on what do you think @rhafer ?